### PR TITLE
IGNITE-20852 Opposite connection attempts may cause connection failure

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-20852

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/message/HandshakeRejectionReason.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    STALE_LAUNCH_ID,
+    /** The sender has detected a clinch and decided to terminate this handshake in favor of the competitor. */
+    CLINCH;
+
+    /**
+     * Returns {@code true} iff the rejection is not expected and should be treated as a critical failure (requiring
+     * the rejected node to restart).
+     */
+    public boolean critical() {
```
Comment (sergey-chugunov-1985) [inline modules/network/src/main/java/org/apache/ignite/internal/network/recovery/message/HandshakeRejectionReason.java:39]: How about renaming `critical` to something like `hazardous` to make it clearer that we'd better send this node into the FailureHandler mouth?

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java [hunk 2]
```diff
@@ -237,70 +233,131 @@ private void onHandshakeStartMessage(HandshakeStartMessage message) {
...
 
-        sendHandshakeRejectedMessage(rejectionMessage, reason);
+        sendHandshakeRejectedMessage(rejectionMessage, message);
+    }
+
+    private void onHandshakeRejectedMessage(HandshakeRejectedMessage msg) {
+        boolean ignorable = stopping.get() || !msg.reason().critical();
+
+        if (ignorable) {
+            LOG.debug("Handshake rejected by server: {}", msg.message());
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java:292]: Inside, there is a check, so if debug level is not enabled, nothing will be logged. If we add the check here, we'll save one method call to `message()` (negligible) and one allocation to create a vararg array. This saving seems to be not important as this code is not hot, we don't handle a million handshakes per second. But we'll have to pay with one line for this. I'm not sure it's worth it.

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryDescriptor.java [hunk 3]
```diff
@@ -138,36 +140,49 @@ public String toString() {
...
+                return null;
+            }
+
+            return acquiry;
+        });
+
+        if (oldAcquiry != null && oldAcquiry.channel() == ctx.channel()) {
+            // We have successfully released the descriptor.
+            // Let's mark the clinch resolved just in case.
+            oldAcquiry.markClinchResolved();
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryDescriptor.java:154]: An acquiry is a local thing, and a clinch is a distributed state, so we cannot see whether there is (was) a clinch or not.

File: modules/network/src/main/java/org/apache/ignite/internal/network/handshake/HandshakeManager.java [hunk 4]
```diff
@@ -48,9 +49,20 @@ default void onConnectionOpen() {
...
-    CompletableFuture<NettySender> handshakeFuture();
+    CompletableFuture<NettySender> localHandshakeFuture();
+
+    /**
+     * Returns final future that represents the handshake operation. This represents completion of either
+     * current handshake or the inverse handshake if it wins (and the current one loses).
+     *
+     * @return Final future that represents the handshake operation.
+     */
+    CompletionStage<NettySender> finalHandshakeFuture();
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/handshake/HandshakeManager.java:67]: 'final' says that this is the result we are interested in. 'ultimate' seems to be ok as well, but it seems too loud.

'global' would be about a different property: not about the 'final result we want to obtain', but that it's common for everyone, and this is not true (another side would have its own future).

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java [hunk 5]
```diff
@@ -74,7 +76,13 @@ public class RecoveryClientHandshakeManager implements HandshakeManager {
...
 
     /** Handshake completion future. */
-    private final CompletableFuture<NettySender> handshakeCompleteFuture = new CompletableFuture<>();
+    private final CompletableFuture<NettySender> localHandshakeCompleteFuture = new CompletableFuture<>();
+
+    /**
+     * Master future used to complete the handshake either with the results of this handshake of the competing one
+     * (in the opposite direction), if it wins.
+     */
+    private final CompletableFuture<CompletionStage<NettySender>> masterHandshakeCompleteFuture = new CompletableFuture<>();
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java]: Changed it to `resulting`

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java [hunk 6]
```diff
@@ -237,70 +233,131 @@ private void onHandshakeStartMessage(HandshakeStartMessage message) {
...
+
+    private void giveUpClinch() {
+        RecoveryDescriptor descriptor = recoveryDescriptorProvider.getRecoveryDescriptor(
+                remoteConsistentId,
+                remoteLaunchId,
+                connectionId
+        );
+
+        DescriptorAcquiry myAcquiry = descriptor.holder();
+        assert myAcquiry != null;
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java:316]: We cannot do anything here 'easily', and the protocol seems to be designed around trust to the other side. If this has to be changed, we'll redesign the protocol, but I think this should be solved by other means (firewall and TLS auth)

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/message/HandshakeRejectionReason.java (hunk 1 above)
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/recovery/message/HandshakeRejectionReason.java]: Renamed it to `restartRequired()`

**Review by sergey-chugunov-1985 — APPROVED** [review]

**Pull request closed without merging**